### ISSUE TITLE
How to train a model for 2 or more people?

### ISSUE THREAD

**Issue by AntouanK** [victorchall/EveryDream-trainer#22]
The guide you have ([MICROMODELS.MD](https://github.com/victorchall/EveryDream-trainer/blob/main/doc/MICROMODELS.MD)) works fine when I try a base model ( SD1.5 for example ) and one person.

I use a script like this : 
```
#!/bin/bash
#

BASE_MODEL=/.../SD1.5.ckpt
PROMPT=sks
PROMPT_TRAINING_DIR=/.../Pictures/sks/clean

python main.py \
    --base configs/stable-diffusion/v1-finetune_micro.yaml \
    -t \
    --actual_resume $BASE_MODEL \
    -n $PROMPT \
    --gpus 0, \
    --data_root $PROMPT_TRAINING_DIR
```

where `sks` is the person I want, and in the `/sks/clean` I have good face photos of that person in 512x512.

the problem is, if I take the output model, and put is as "BASE_MODEL" and try to train for another person, the results are weird.
It kind of knows the one person, but the second is a mix of the two!!

Also, another issue I saw is that if I take some existing trained models from others ( let's say this one https://huggingface.co/wavymulder/Analog-Diffusion ) the results are not good at all. I cannot get that sks person to appear anything close to what it is.
I'm wondering what the issue is.
When a model is re-trained with new photos, it changes so much that I cannot use it as a base for another one?
Or am I missing something?

**Comment by noprompt** [comment]
@AntouanK Train both people at the same time.

As for training on top of an existing model, if you cannot get access to the training set used, try generating a hundred good images from the model and save them with their prompts (since they will contain the keywords recognized by the original model ie. `analog style`). This may not work perfectly but it should work better than training right on top.

IMO it's best to train with `v1-finetune_everydream` and always keep a set of LAION images in your training set. I like to use 1000 randomly selected LAION images with their captions per session. For example:

```
training_samples/
  subject_1/
  ...
  subject_n/
  laion/
```

**Comment by AntouanK** [comment]
@noprompt 

> Train both people at the same time.

How would I do that?
Can you give examples of the command and what images/prompt I need?

**Comment by noprompt** [comment]
@AntouanK The data set organization is documented on the README [here](https://github.com/victorchall/EveryDream-trainer#data-set-organization) and then you would train using the command documented [here](https://github.com/victorchall/EveryDream-trainer#starting-training). So your `training_samples` folder should look like

```
/training_samples/sks
/training_samples/sks/sks1
/training_samples/sks/sks2
/training_samples/sks/laion
```

Where `sks1` has pictures of the first person, `sks2` has pictures of the second person, and laion has pictures from LAION. The command you will run will look like

```
BASE_MODEL=/.../SD1.5.ckpt
PROMPT=sks
PROMPT_TRAINING_DIR=./training_samples/sks

python main.py \
    --base configs/stable-diffusion/v1-finetune_micro.yaml \
    -t \
    --actual_resume $BASE_MODEL \
    -n $PROMPT \
    --gpus 0, \
    --data_root $PROMPT_TRAINING_DIR
```

**Comment by AntouanK** [comment]
Thank you for the example.
I'm a bit confused though.
The way I did it so far is replace the `sks` with a name, like `johndoe` or `joedoe` and then give that person's photos.
So, I don't get what `sks`, `sks1` and `sks2` stand for here.
Maybe I misunderstood something.

**Comment by noprompt** [comment]
@AntouanK Your goal is to train two people, yes? You trained one person to get a model which you then used as a base to train another person and that didn't work.

> the problem is, if I take the output model, and put is as "BASE_MODEL" and try to train for another person, the results are weird.
It kind of knows the one person, but the second is a mix of the two!!

If you want to train two people, you should train them at the same time to avoid these kinds of results and use the `v1-finetune_everydream.yaml` configuration. **Do not use the micro workflow.**

> The way I did it so far is replace the `sks` with a name, like `johndoe` or `joedoe` and then give that person's photos.
> So, I don't get what `sks`, `sks1` and `sks2` stand for here.

I was using the name `sks` you provided for illustration. Substituting the names `johndoe` and `joedoe` your training folder should look like

```
/project-name
/project-name/johndoe
/project-name/joedoe
/project-name/laion
```

Inside `johndoe` and `joedoe` you should have either image files where the names of the files are the captions for the picture

```
a_picture_of_john_doe_at_the_beach.jpg
a_picture_of_john_doe_smiling.jpg
...
```

or pairs of image files and text files like
```
john_doe_01.jpg
john_doe_01.txt
...
```

where the `.txt` files contain the caption for that image ie. `a picture of john_doe_smiling`. We do the same for the LAION directory (use the LAION scraping script in the other repository as mentioned in the README to get these images.)

I could be wrong but judging from the variable you used in your shell `PROMPT`, I think you are confused about how the fine tuning works. It is not like DreamBooth where you have a directory of images all associated with a single keyword or phrase. It is much more general where the image caption are used in conjunction with the image to produce a richer model; this method of training is more like how SD was trained. If it helps to make things more clear, replace the shell variable `PROMPT` with  `PROJECT_NAME`.

To tie this all together, lets say we're going to fine tune a model on two people: John Doe, and Joe Doe. Then, in the `training_samples` directory we will create a new folder called `doe` and place in there our files as I just discussed.

```
/doe
/doe/johndoe/a_picture_of_john_doe_at_the_beach.jpg
/doe/johndoe/john_doe_wearing_t_shirt_and_jeans.jpg
...
/doe/joedoe/a_picture_of_joe_doe_smiling.jpg
/doe/joedoe/joe_doe_wearing_sunglasses.jpg
...
/doe/laion
...
```

Now, to train fine tune a model that understands both John Doe _and_ Joe Doe robustly

```
BASE_MODEL=/.../SD1.5.ckpt
PROJECT=sks
PROJECT_TRAINING_DIR=./training_samples/doe

python main.py \
    --base configs/stable-diffusion/v1-finetune_everydream.yaml \
    -t \
    --actual_resume $BASE_MODEL \
    -n $PROJECT \
    --gpus 0, \
    --data_root $PROJECT_TRAINING_DIR
```

Once the training has completed and you prune the model, you should be able generate images of John Doe and Joe Doe without your previous issues.

I hope this helps. If you still don't understand, reproduce the workflow I have described here for your project and try it for yourself.

**Comment by AntouanK** [comment]
Thank you so much for the detailed explanation. :+1: 

I'll try to make new sets of captioned images and give it a try.

**Comment by victorchall** [comment]
Think your questions are covered above, you can train as many "things" as you want simply by using the captions, and you can capture not just the "thing" but the context.

The prune script is run this way:

python scripts/autoprune_all.py

This will find all the ckpts in the logs folder, copy them down to the root folder, and "prune" them from 11gb down to 2gb, then you can go use them with whatever inference/webui program you want.  You can choose to delete the 11gb files, or keep the last one if you wish to resume training on it (if you find it is undertrained, just point --actual_resume to it)

**Comment by victorchall** [comment]
SD2.x models will be supported in ED2.0 which is pending release (soon(tm)

**Comment by victorchall** [comment]
Think that covers the original topic, if you have more question LMK.  I usually faster to respond on the discord as well.

**Comment by AntouanK** [comment]
ok, great.

the problem I have now is that I get python install errors.
I'm totally clueless on python, and by checking out all those AI repos on my machine, I think I have a big mess and I don't know how to "reset" it.
Some have conda environments, some wanted pip install, etc.
( any specific reason why this repo has `environment.yaml` instead of `environment.yml`? )
I also somehow have both python 3.9 and 3.10 on my machine. :man_shrugging: 
Very messy. I tried to delete anaconda completely and restart, but I get errors again.

For example, right now I get this
![image](https://user-images.githubusercontent.com/4569111/208289329-06ecb479-4034-49a8-a12f-6819b14f3c39.png)

when I run
```
conda env create -f environment.yaml
```
so, I'm stuck trying to get that part done.
( it had worked before fine, so I know it's because of other repos I installed with pip/conda or something :( )

**Comment by AntouanK** [comment]
the solution I found so far is the tool I know best.
Docker :grimacing: 

I made a docker container based on `continuumio/anaconda3` and cloned the repo in there.
Conda create / activate works fine in there ( after messing with bash a bit )
I'm running the training now, I hope it works. :pray: :crossed_fingers: 


( also I noticed last time I used it that it some time crashes my linux desktop environment and I have to re-login and start everything from scratch. not sure how to prevent that from happening )

**Comment by AntouanK** [comment]
false hopes :joy: 

I get this error.
![image](https://user-images.githubusercontent.com/4569111/208297845-9e2b7e5c-5604-401b-a222-75882599b6ef.png)

I think this is related to SD 2.1.
That's why I asked earlier if I can use that script to train with 2.x.
I guess I should have read [this answer](https://github.com/victorchall/EveryDream-trainer/issues/22#issuecomment-1356506555). :man_facepalming: 

I'll try again with 1.5

**Comment by AntouanK** [comment]
That worked great with as 1.5 based model. :tada: 

I'll make a repo and checkout my files, in case anyone else wants to use a dockerized version.
I think I'll stick with that. Local conda/pip/whatever packages are a nightmare if you don't know about python :( 

Waiting on the 2.0 release :smiley:

**Comment by AntouanK** [comment]
training works fine.
the models I get are not good at all though.
meaning that they don't learn how the people look. they look vaguely similar but definitely not the same. more like distant cousins :joy: 

I guess either my photos are bad, or the cropping/captioning is bad. :/
